golang/cosmos/daemon/cmd/root.go

```diff
@@ -263,6 +281,57 @@ func (ac appCreator) newApp(
 	)
 }
 
+const (
+	FlagExportDir                   = "export-dir"
+	ExportedGenesisFileName         = "genesis.json"
+	ExportedSwingStoreDirectoryName = "swing-store"
+)
+
+func extendCosmosExportCommand(cmd *cobra.Command) {
+	cmd.Flags().String(FlagExportDir, "", "The directory where to create the genesis export")
+	err := cmd.MarkFlagRequired(FlagExportDir)
+	if err != nil {
+		panic(err)
+	}
+
+	originalRunE := cmd.RunE
+
+	extendedRunE := func(cmd *cobra.Command, args []string) error {
+		serverCtx := server.GetServerContextFromCmd(cmd)
+
+		exportDir, _ := cmd.Flags().GetString(FlagExportDir)
+		err := os.MkdirAll(exportDir, os.ModePerm)
+		if err != nil {
+			return err
+		}
+
+		genesisPath := filepath.Join(exportDir, ExportedGenesisFileName)
+		swingStoreExportPath := filepath.Join(exportDir, ExportedSwingStoreDirectoryName)
+
+		err = os.MkdirAll(swingStoreExportPath, os.ModePerm)
+		if err != nil {
+			return err
+		}
+		serverCtx.Viper.Set(gaia.FlagSwingStoreExportDir, swingStoreExportPath)
+
+		genesisFile, err := os.OpenFile(genesisPath, os.O_CREATE|os.O_EXCL|os.O_WRONLY, os.ModePerm)
+		if err != nil {
+			return err
+		}
+		defer genesisFile.Close()
+
+		cmd.SetOut(genesisFile)
+
+		return originalRunE(cmd, args)
+	}
+
+	// OnExportHook is only set when the VM is not started
+	// We don't need to override the behavior if the process will exec to the VM
+	if OnExportHook == nil {
+		cmd.RunE = extendedRunE
+	}
```

Comment:
I don't understand why this special case is necessary.  Isn't it more robust to have the same behaviour regardless of whether or not we're starting the VM?

```suggestion
	cmd.RunE = extendedRunE
```

Reply:
Without it we'd try to create the genesis file twice, and fail the second time around. I initially worked around that by truncating existing files

Reply:
Ah, okay.  That's more subtle than I saw at first.  I suggest instead detecting whether `sender` has been set, and only running this function in that particular case, as I have commented above in this file.

Then, you can omit this condition based on `OnExportHook`, which is a separate concern from whether the current binary is executing a different one to obtain the controller (`sender == nil`).

Reply:
I did a variation on this as I think we should fail early if `--export-dir` is missing. PTAL